### ISSUE TITLE
error sending message '[Title]': remote error: tls: user canceled

### ISSUE THREAD

**Issue by kulterryan** [knadh/listmonk#1717]
**Version:**
 - listmonk: [eg: v1.0.0]
 - OS:Ubuntu/Railway

**Description of the bug and steps to reproduce:**
Getting TLS error after a specific time interval when logged in the application.

**Screenshots:**
![image](https://github.com/knadh/listmonk/assets/36154235/fdf7a0f7-2235-4e16-a175-6c587c475a11)

**Comment by knadh** [comment]
hm, sounds like an an SMTP/TLS configuration issue. What SMTP provider are you using?

**Comment by kulterryan** [comment]
I'm using AWS SES

**Comment by knadh** [comment]
Unable to reproduce this with SES in production. Could've something to do with network configuration.

**Comment by MaximilianKohler** [comment]
I just started getting this with SES as well on v2.5.1. I've been sending 2-3k emails per day without issue and gradually increased it to 50k, and it happened for the first time now even though my SES limit is 500k.

```
2024/03/21 08:52:00 start processing campaign (<campaign name> 2024-03-21 50k)
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968723: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968726: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968724: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968725: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968728: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968727: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968729: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968730: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968731: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968732: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968734: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968733: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968735: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968737: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968739: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968740: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968741: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968743: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968747: remote  error: tls: user canceled
2024/03/21 09:12:01 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 968748: remote  error: tls: user canceled
2024/03/21 09:23:47 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 983818: 421 Connection closed by server. Too many messages in one session
2024/03/21 09:23:48 error sending message in campaign <campaign name> 2024-03-21 50k: subscriber 983829: remote  error: tls: user canceled
2024/03/21 09:31:23 campaign (<campaign name> 2024-03-21 50k) finished
```

Are you saying this has to do with my server provider (Hetzner)?

![Screenshot 2024-03-21 101551](https://github.com/knadh/listmonk/assets/3606996/896151d5-9000-4091-a294-8013b4e6e7d4)
![Screenshot 2024-03-21 101613](https://github.com/knadh/listmonk/assets/3606996/d8fc7992-0ba9-4c92-8bd8-8e63f562b872)
![Screenshot 2024-03-21 101626](https://github.com/knadh/listmonk/assets/3606996/7aea91ce-f3ea-4430-98d9-7fd7069c77c3)
![Screenshot 2024-03-21 103859](https://github.com/knadh/listmonk/assets/3606996/51a2cf7c-c87e-4ba9-ba6e-6615012beb36)

**Comment by MaximilianKohler** [comment]
I tried lowering the send rate from 10x10 to 10x7 and then 10x5 and it still happens. I thought it might be the recipient email server (gmail, hotmail, etc.) rejecting my emails, but looking at the subscribers it failed for, they're very diverse. IE: they're not all from gmail or hotmail, but many providers.

After lowering it to 10x5 it didn't take long for it to start giving errors, and then the errors occurred sporadically throughout:

```
2024/03/24 08:58:04 start processing campaign (<campaign name>)
2024/03/24 09:18:04 error sending message in campaign <campaign name>: subscriber 359952: 421 Connection closed by server. Too many messages in one session
```

So from 8:58 to 9:18 it couldn't have even sent that many messages at 10x5.

**Comment by MaximilianKohler** [comment]
I found this https://stackoverflow.com/questions/49111453/sending-email-with-aws-ses-via-smtp-error-421 which says it might be solved by changing the port. Here's another discussion on that: https://github.com/knadh/listmonk/issues/1615

I switched to STARTTLS 587 and got the exact same errors.

**Comment by MaximilianKohler** [comment]
@knadh An Amazon expert replied: https://repost.aws/questions/QULOBzwPLMS2GkrUgOjWNshQ/amazon-ses-remote-error-tls-user-canceled-421-connection-closed-by-server-too-many-messages-in-one-session

> Hi there
> 
> Can you provide some details on how your application is sending mail? Is it using some sort of keep alive to hold a connection open to the SMTP server? If so, you might want to try limiting that. Also, do you have any retry/backoff in your application code?
> 
> See https://docs.aws.amazon.com/ses/latest/dg/troubleshoot-smtp.html#troubleshoot-smtp-response-codes
> 
> >    You should retry SMTP requests that receive 400 errors. We recommend that you implement a system that retries requests with progressively longer wait times (for example, wait 5 seconds before retrying, then wait 10 seconds, and then wait 30 seconds). If the third retry doesn't succeed, wait 20 minutes, and then repeat the process. To see an example of an implementation that uses an exponential retry policy, see How to handle a "Throttling - Maximum sending rate exceeded" error on the AWS Messaging and Targeting Blog.

**Comment by MaximilianKohler** [comment]
Hetzner says it's not a limitation on their end.

So the solution seems to be increasing the default "retries" from 2 to 4 under `/admin/settings -> SMTP`, but only if there's a progressive delay implemented as the Amazon guy said.

Does listmonk have a progressive delay? If not, that would be the solution no?

**Comment by MaximilianKohler** [comment]
I think this needs to be reopened. I changed the retries from 2 to 5 and I still get the errors with as small a list as 4000 subscribers.

```
2024/10/25 09:02:31 error sending message in campaign xxx: 451 4.4.2 Timeout waiting for data from client.
2024/10/25 09:02:31 error sending message in campaign xxx: remote error: tls: user canceled
```

**Comment by kulterryan** [comment]
Hi @MaximilianKohler 
Did you tried upgrading to latest version?

**Comment by MaximilianKohler** [comment]
I'm on v2.5.1. I don't recall anything in v3.0 or 4.0 addressing this issue. I'm running listmonk on a Hetzner server. Almalinux 9, binary.

**Comment by MaximilianKohler** [comment]
I changed:
Batch size 2k from 5k, Concurrency & rate from 10x10 to 6x6, retries 5 to 3, wait timeout 5s to 15s.


And I still got a bunch of the same errors.

```
421 Connection closed by server. Too many messages in one session
remote error: tls: user canceled
```
I then changed "Max. connections. Maximum concurrent connections to the server." from 10 to 5. And increased Concurrency & rate from 6x6 to 8x8. 

Still happened. So there doesn't seem to be anything I can change in the settings to fix this. The only solution seems to be increasing the delay between retries to something more significant, per what the Amazon dev said: https://github.com/knadh/listmonk/issues/1717#issuecomment-2040780489. Retrying them at the end of the campaign seems like it could work.

Also, `Idle timeout` and `Wait timeout` have the exact same help text, so it's not clear how they differentiate.
